tutorials/provisional/mosaic.py

```diff
@@ -0,0 +1,285 @@
+labels used to lay out the Figure.  By returning data structures with
+names, it is easier to write plotting code that is independent of the
+Figure layout.
+
+
+This is inspired by a `proposed MEP
+<https://github.com/matplotlib/matplotlib/pull/4384>`__ and the
+`patchwork <https://github.com/thomasp85/patchwork>`__ library for R.
+While we do not implement the operator overloading style, we do
+provide a Pythonic API for specifying (nested) Axes layouts.
+
+"""
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+# Helper function used for visualization in the following examples
+def identify_axes(ax_dict, fontsize=48):
+    """
+    Helper to identify the Axes in the examples below.
+
+    Draws the label in a large font in the center of the Axes.
+
+    Parameters
+    ----------
+    ax_dict : Dict[str, Axes]
+        Mapping between the title / label and the Axes.
+
+    fontsize : int, optional
+        How big the label should be
+    """
+    kw = dict(ha='center', va='center', fontsize=fontsize, color='darkgrey')
+    for k, ax in ax_dict.items():
+        ax.text(0.5, 0.5, k, **kw)
+
+
+###############################################################################
+# String specification
+# ====================
+#
+# Using single letters we can "draw" the Axes we want.  The following
+
+
+layout = """
+    AB
+    CD
+    """
+
+###############################################################################
+# will give us 4 Axes laid out in a 2x2 grid and generates the
+# same figure layout as ``fig.subplots(2, 2)``.
+
+fig = plt.figure(constrained_layout=True)
+ax_dict = fig.subplot_mosaic(layout)
+identify_axes(ax_dict)
+
+###############################################################################
+# A key difference between `.Figure.subplots` and
+# `.Figure.subplot_mosaic` is the return value. While
+# the former returns an array for index access, the
+# latter returns a dictionary mapping the labels to the
+# `.axes.Axes` instances created
+
+print(ax_dict)
+
+###############################################################################
+# For comparison we could do:
+
+fig2 = plt.figure(constrained_layout=True)
+ax_array = fig2.subplots(2, 2, squeeze=False)
+identify_axes({str((j, k)): ax
+               for j, row in enumerate(ax_array)
+               for k, ax in enumerate(row)})
+
+###############################################################################
+# If we wanted to use the upper left Axes we would do either
+
+upper_left = ax_dict['A']
+upper_left = ax_array[0, 0]
+
+
+###############################################################################
+# Something else we can easily do with `.Figure.subplot_mosaic` is
+# specify a span.  If we want to add an Axes to the right side of
+# our Figure that spans the full vertical space we could do
+
+axd = plt.figure(constrained_layout=True).subplot_mosaic(
+    """
+    ABE
+    CCE
+    """
+)
+identify_axes(axd)
```

Comment:
Keep ABCD and just run the same code as above

Reply:
I think there is value in changing the letters up make clear that there is no actual meaning in the values of the letters.

Reply:
On more thought, we make that point later on, will change this to use D.